clang/lib/Driver/Driver.cpp

```diff
@@ -8457,9 +8457,17 @@ class ToolSelector final {
     return dyn_cast<JobAction>(CurAction);
   }
 
+  // We need to collapse the separate compilation steps when performing
+  // a third party host compilation step for SYCL offloading.  We don't know
+  // what the third party compiler is capable of, so only allow for object
+  // creation when performing -save-temps.
+  bool SYCLHostCompiler =
+      BaseAction->isHostOffloading(Action::OFK_SYCL) &&
+      C.getArgs().hasArg(options::OPT_fsycl_host_compiler_EQ);
+
   /// Return true if an assemble action can be collapsed.
   bool canCollapseAssembleAction() const {
-    return TC.useIntegratedAs() && !SaveTemps &&
+    return TC.useIntegratedAs() && !(SaveTemps && !SYCLHostCompiler) &&
```

Comment:
nit: I find this a bit hard to interpret.
We want to collapse/combine the compilation steps when `SYCLHostCompiler` is true and also when `SaveTemps` is true correct?

Reply:
The scenarios when we want to collapse:
 - default compilation && !save-temps
 - fsycl-host-compiler && save-temps (host compilation steps only)
Effectively maintaining the same behaviors for prohibiting collapsing as before, but adding the -fsycl-host-compiler exception.